unit-test-security-rules-v9/test/firestore.spec.mjs

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import assert from 'assert';
+import { readFileSync } from 'fs';
+
+// Workarounds for npm not recognizing named exports in cjs.
+import appNs from 'firebase/app';
+const { initializeApp, deleteApp, getApps } = appNs;
+
+// We can use firestore/lite since we don't need persistence or use listeners.
+// Feel free to switch to firebase/firestore if your tests use onSnapshot etc.
+import firestoreNs from 'firebase/firestore/lite';
+const { getFirestore, useFirestoreEmulator, doc, getDoc, setDoc, serverTimestamp, setLogLevel } = firestoreNs;
+
+import adminAppNs from 'firebase-admin/app';
+const { initializeApp: initializeAdminApp } = adminAppNs;
+import adminFirestoreNs from 'firebase-admin/firestore';
+const { getFirestore: getAdminFirestore } = adminFirestoreNs;
+import adminSecurityRulesNs from 'firebase-admin/security-rules';
+const { getSecurityRules } = adminSecurityRulesNs;
+
+/**
+ * The emulator will accept any project ID for testing.
+ */
+const PROJECT_ID = "demo-firestore-emulator";
+
+/**
+ * The FIRESTORE_EMULATOR_HOST environment variable is set automatically
+ * by "firebase emulators:exec"
+ */
+const { FIRESTORE_EMULATOR_HOST } = process.env;
+if (!FIRESTORE_EMULATOR_HOST) {
+  throw new Error('Missing environment variable FIRESTORE_EMULATOR_HOST. Consider running tests via `firebase emulators:exec "npm test"`.');
+}
+
+const { hostname: FIRESTORE_HOST, port: FIRESTORE_PORT } = new URL(`http://${FIRESTORE_EMULATOR_HOST}`);
+
+before(async () => {
+  initializeAdminApp({ projectId: PROJECT_ID });
+
+  // Load the rules file before the tests begin.
+  // TODO: Uncomment once it works.
+  // getSecurityRules()
```

Comment:
This really scares me since one missing env var and you've changed rules in prod. I think an emulator-only utility here has big value.

Reply:
Same. I've tried to enforce the env var on the top of the file, but I think it's debatable. I'm still worried about developers copying only one snippet (this one) and get surprising results.